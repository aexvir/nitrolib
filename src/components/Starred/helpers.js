// @flow
import isAfter from "date-fns/isAfter"; // keys
import * as R from "ramda";

import type { StarredItem, PassengersCount } from "../../records/Starred";
import type { Journey } from "../../records/Journey";
import { TRANS_KEY_SINGLE, TRANS_KEY_MULTI } from "./consts";

export const getSum = (object: PassengersCount) =>
  Object.keys(object).reduce((previous, key) => previous + object[key], 0);

export const isMulti = (object: PassengersCount) =>
  Object.keys(object).filter(key => object[key] > 0).length > 1;

export const getTransKey = (object: PassengersCount) => {
  if (isMulti(object)) {
    return TRANS_KEY_MULTI;
  }

  const keyWithValue = Object.keys(object).find(key => object[key] > 0) || "adults";
  const source = TRANS_KEY_SINGLE[keyWithValue];

  return Object.keys(object).length > 1 ? source.many : source.one;
};

export const getPrice = (journey: Journey) => R.prop("default", journey.prices);

export const getBestPrice = (journey: Journey) => {
  const { providers } = journey;
  const prices = providers.map(trip => trip.price);

  if (R.isEmpty(providers)) {
    return getPrice(journey);
  }

  return prices.length > 1 ? Math.min(...prices) : prices[0];
};

export const itemIsValid = (item: StarredItem) =>
  isAfter(R.prop("utc", item.journey.trips[0].flights[0].departure.when), Date.now());
