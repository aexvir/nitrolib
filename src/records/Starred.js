// @flow
import type { Journey } from "./Journey";

export type CabinClass = "economy" | "business" | "first" | "premium";

export type PassengersCount = {|
  adults: number,
  children: number,
  infants: number,
|};

export type StarredFormData = {|
  origin: string,
  destination: string,
  outboundDate: string,
  inboundDate: string,
  multicity: string,
  salesman: string,
  passengers: PassengersCount,
  cabinClass: CabinClass,
  filters: any,
  lang: string,
  places: Array<{ id: string, slug: string }>,
  returnUrl: string,
  starType: string,
|};

export type StarredItem = {|
  id: string,
  form: StarredFormData,
  lastPrice: number,
  journey: Journey,
  priceUpdatedAt: ?Date,
  createdAt: Date,
  updatedAt: Date,
|};
