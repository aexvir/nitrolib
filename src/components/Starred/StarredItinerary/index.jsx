// @flow
import * as React from "react";
import styled, { css } from "styled-components";
import Star from "@kiwicom/orbit-components/lib/icons/StarFull";
import Share from "@kiwicom/orbit-components/lib/icons/Share";

import TimeInWords from "../../DistanceInWords";
import Price from "../../Price";
import Time from "../../Time";
import Translate from "../../Translate";
import TranslateNode from "../../TranslateNode";
import StarredSegment from "../StarredTripSegment";
import mq from "../../../styles/mq";
import { getTransKey, getBestPrice } from "../helpers";
import type { ThemeProps } from "../../../records/Theme";
import type { Journey } from "../../../records/Journey";
import type { PassengersCount } from "../../../records/Starred";
import { themeDefault } from "../../../records/Theme";
import { PASSENGERS_COUNT } from "../consts";

type Props = {|
  updated: Date,
  journey: Journey,
  passengerCount: number,
  passengers: PassengersCount,
  passengerMulty: boolean,
  price: number,
  onRemove: () => void,
  priceUpdatedAt: Date,
  isValid: boolean,
  created: string,
|};

const Added = styled.div`
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkLight};
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextSmall};
`;

Added.defaultProps = {
  theme: themeDefault,
};

const Info = styled.div`
  display: flex;
  min-width: 80px;
  width: 80px;
  margin-right: 5px;
  flex-direction: column;
  ${mq.ltBigMobile(css`
    margin-top: 10px;
  `)};
`;

const FlightPrice = styled.div`
  font-weight: ${({ theme }: ThemeProps) => theme.orbit.fontWeightBold};
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkDark};
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextLarge};
`;

FlightPrice.defaultProps = {
  theme: themeDefault,
};

const PriceInfo = styled.div`
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkLight};
  text-transform: lowercase;
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextSmall};
  line-height: 22px;
  max-width: 80px;
  text-overflow: ellipsis;
  overflow: hidden;
`;

PriceInfo.defaultProps = {
  theme: themeDefault,
};

const Wrapper = styled.div`
  padding: 15px 20px;
  cursor: pointer;
  display: flex;
  flex-direction: column;
  box-shadow: rgb(232, 237, 241) 0px -1px inset;
  background: ${({ theme }) => theme.orbit.paletteWhite};
  &:hover {
    background: ${({ theme }) => theme.orbit.paletteWhiteHover};
  }
`;

Wrapper.defaultProps = {
  theme: themeDefault,
};

const Flights = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
`;

const WrapperInner = styled.div`
  display: flex;
  margin-top: 15px;
  align-items: center;
  ${mq.ltBigMobile(css`
    flex-direction: column-reverse;
    align-items: flex-start;
  `)};
`;

const ActionButtons = styled.div`
  display: flex;
  justify-content: space-between;
  max-width: 55px;
  margin-top: 5px;
`;

const RemoveButton = styled.div``;

const ShareIcon = styled(Share)`
  &:hover {
    color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkLight};
  }
`;

ShareIcon.defaultProps = {
  theme: themeDefault,
};

const StarredItinerary = ({
  updated,
  journey,
  price,
  onRemove,
  isValid,
  created,
  passengers,
  priceUpdatedAt,
  passengerCount,
  passengerMulty,
}: Props) => {
  const lastUpdate = <Time time={updated} />;
  const getPriceUpdated = () => {
    if (!isValid) {
      return (
        <div>
          <Translate t={__("starred.flight_not_available")} />
          <RemoveButton onClick={onRemove}>
            <Translate t={__("starred.remove_starred")} />
          </RemoveButton>
        </div>
      );
    }
    if (price !== getBestPrice(journey)) {
      return (
        <Translate
          t={__("starred.price_update_changed")}
          values={{
            priceUpdatedAt,
            origPrice: price,
          }}
        />
      );
    }
    return priceUpdatedAt ? (
      <TranslateNode t={__("starred.price-update")} values={{ lastUpdate }} />
    ) : (
      <TranslateNode
        t={__("starred.created_at")}
        values={{ createdAt: <TimeInWords time={created} /> }}
      />
    );
  };

  return (
    <Wrapper>
      <Added>{getPriceUpdated()}</Added>
      <WrapperInner>
        <Info>
          <FlightPrice>
            <Price value={getBestPrice(journey)} />
          </FlightPrice>
          <PriceInfo>
            {passengerCount > 1 &&
              (passengerMulty ? (
                <Translate t={__("result.total_price")} />
              ) : (
                <TranslateNode
                  t={PASSENGERS_COUNT[getTransKey(passengers)]}
                  values={{ count: passengerCount }}
                />
              ))}
          </PriceInfo>
          <ActionButtons>
            <RemoveButton onClick={onRemove}>
              <Star color="warning" />
            </RemoveButton>
            <ShareIcon color="tertiary" />
          </ActionButtons>
        </Info>
        <Flights>
          {journey.trips.map(trip => {
            const { flights } = trip;
            const getDeparture = flights[0].departure;
            const getArrival = flights[flights.length - 1].arrival;
            return (
              <StarredSegment
                key={trip.flights[0].id}
                departure={getDeparture}
                arrival={getArrival}
              />
            );
          })}
        </Flights>
      </WrapperInner>
    </Wrapper>
  );
};

export default StarredItinerary;
