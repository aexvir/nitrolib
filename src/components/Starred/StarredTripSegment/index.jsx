// @flow
import * as React from "react";
import styled from "styled-components";
import FlightDirect from "@kiwicom/orbit-components/lib/icons/FlightDirect";

import type { DepartureArrival } from "../../../records/Flight";
import type { ThemeProps } from "../../../records/Theme";
import { themeDefault } from "../../../records/Theme";
import Flex from "../../../primitives/Flex";
import Day from "../../Day";
import Duration from "../../Duration";

type Props = {|
  departure: DepartureArrival,
  arrival: DepartureArrival,
|};

const Wrapper = styled.div`
  margin: 5px 0 5px 0;
  position: relative;
`;

const TripDate = styled.div`
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkLight};
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextSmall};
  padding-top: 5px;
`;

TripDate.defaultProps = {
  theme: themeDefault,
};

const DurationWrapper = styled.div`
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextNormal};
  font-weight: ${({ theme }: ThemeProps) => theme.orbit.fontWeightMedium};
`;

DurationWrapper.defaultProps = {
  theme: themeDefault,
};

const Destination = styled.div`
  display: flex;
  align-items: center;
  height: 16px;
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextNormal};
`;

Destination.defaultProps = {
  theme: themeDefault,
};

const Flight = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  width: 100%;
  margin-right: 10px;
  white-space: nowrap;
`;

const FlightandCode = styled.span`
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
`;

const FlightTime = styled.div`
  display: flex;
  align-items: center;
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkLight};
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextSmall};
  white-space: nowrap;
  padding-top: 5px;
`;

const LongText = styled.div`
  position: absolute;
  top: -16px;
  left: 0;
  right: 0;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const WrapperTime = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  min-width: 80px;
`;

FlightTime.defaultProps = {
  theme: themeDefault,
};

const StarredTripSegment = ({ departure, arrival }: Props) => (
  <Wrapper>
    <Flex x="space-between" y="center">
      <Flight>
        <LongText>
          <Destination>
            <FlightandCode>
              {departure.where.name} ({departure.where.code})
            </FlightandCode>
            <FlightDirect />
            <FlightandCode>
              {arrival.where.name} ({arrival.where.code})
            </FlightandCode>
          </Destination>
          <TripDate>
            <Day date={departure.when.local} />
          </TripDate>
        </LongText>
      </Flight>
      <WrapperTime>
        <DurationWrapper>
          <Duration timeFrom={departure.when.local} timeTo={arrival.when.local} format="HH:mm" />
        </DurationWrapper>
        <FlightTime>
          <Day date={departure.when.local} format="HH:mm" />
          {" - "}
          <Day date={arrival.when.local} format="HH:mm" />
        </FlightTime>
      </WrapperTime>
    </Flex>
  </Wrapper>
);

export default StarredTripSegment;
