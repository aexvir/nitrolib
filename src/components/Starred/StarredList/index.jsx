// @flow
import * as React from "react";
import styled from "styled-components";

import { getSum, isMulti, itemIsValid } from "../helpers";
import StarredItinerary from "../StarredItinerary";
import type { StarredItem } from "../../../records/Starred";
import Text from "../../Translate";
import { themeDefault } from "../../../records/Theme";
import type { ThemeProps } from "../../../records/Theme";

type Props = {|
  trips: Array<StarredItem>,
  onRemove: (arg: number) => void,
  tripsCount: number,
  onSelectStarred: (item: StarredItem) => void,
|};

const NoFlights = styled.div`
  font-size: ${({ theme }: ThemeProps) => theme.orbit.fontSizeTextNormal};
  padding: ${({ theme }: ThemeProps) => theme.orbit.spaceSmall};
`;

NoFlights.defaultProps = {
  theme: themeDefault,
};

const StarredList = ({ trips, onRemove, tripsCount, onSelectStarred }: Props) => {
  const tripList: React$Node =
    trips &&
    trips.map((trip, index) => {
      const { id, updatedAt, lastPrice, journey, priceUpdatedAt, form, createdAt } = trip;
      const { passengers } = form;
      return (
        <StarredItinerary
          key={id}
          passengerCount={getSum(passengers)}
          passengerMulty={isMulti(passengers)}
          isValid={itemIsValid(trip)}
          passengers={passengers}
          price={lastPrice}
          onSelectStarred={() => onSelectStarred(trip)}
          journey={journey}
          onRemove={() => onRemove(index)}
          created={createdAt}
          updated={updatedAt}
          priceUpdatedAt={priceUpdatedAt}
        />
      );
    });

  return tripsCount >= 1 ? (
    tripList
  ) : (
    <NoFlights>
      <Text t={__("starred.no_flights")} />
    </NoFlights>
  );
};

export default StarredList;
