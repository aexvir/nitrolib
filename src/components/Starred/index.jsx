// @flow strict
import * as React from "react";
import StarIcon from "@kiwicom/orbit-components/lib/icons/StarFull";

import type { StarredItem } from "../../records/Starred";
import Button from "../NavBar/primitives/Button";
import Toggle from "../Toggle";
import ClickOutside from "../ClickOutside";
import Desktop from "../Desktop";
import Mobile from "../Mobile";
import Translate from "../Translate";
import TripsContainer from "../TripsContainer";
import { getItem } from "../../services/storage/storage";
import StarredHeader from "./StarredHeader";
import StarredList from "./StarredList";
import StarredFooter from "./StarredFooter";
// import data from "./data";
import { MAX_TRIPS } from "./consts";

type Props = {|
  onSaveTrip: () => void,
  onRemoveTrip: number => void,
  positionMenuDesktop?: number,
  positionMenuTablet?: number,
|};

type State = {|
  starred: Array<StarredItem>,
|};

class Starred extends React.Component<Props, State> {
  state = {
    starred: [],
  };

  componentDidMount() {
    // TODO: remove after implementation
    // localStorage.setItem("starred", JSON.stringify(data));
    const store = getItem("starred");
    if (store) {
      const Items = JSON.parse(store);
      this.setState({
        starred: Items,
      });
    }
  }

  clearStore = () => {
    localStorage.removeItem("starred");
    this.setState({
      starred: [],
    });
  };

  removeTrip = (index: number) => {
    const { starred } = this.state;
    const store = getItem("starred");

    this.setState({
      starred: starred.filter((_, i) => i !== index),
    });

    const removeItem = store.slice(0, index - 1).concat(store.slice(index, store.length));
    localStorage.setItem("starred", JSON.stringify(removeItem));
  };

  render() {
    const { onSaveTrip, positionMenuDesktop, positionMenuTablet } = this.props;
    const { starred } = this.state;
    const starredList = starred.slice(0, MAX_TRIPS);
    const starredCount = starredList.length;
    const starredFooter = starredCount >= 1 && <StarredFooter tripsCount={starredCount} />;
    return (
      <Toggle>
        {({ open, onToggle }) => (
          <>
            {open && (
              <ClickOutside onClickOutside={onToggle}>
                <TripsContainer
                  header={
                    <StarredHeader onClearStorage={this.clearStore} tripsCount={starredCount} />
                  }
                  footer={starredFooter}
                  positionMenuTablet={positionMenuTablet}
                  positionMenuDesktop={positionMenuDesktop}
                >
                  <StarredList
                    onRemove={this.removeTrip}
                    trips={starredList}
                    tripsCount={starredCount}
                    onSave={onSaveTrip}
                  />
                </TripsContainer>
              </ClickOutside>
            )}
            <Desktop>
              <Button onClick={onToggle} color="secondary">
                <Translate t={__("starred.starred")} />
              </Button>
            </Desktop>
            <Mobile>
              <Button onClick={onToggle} color="secondary">
                <StarIcon color="primary" />
              </Button>
            </Mobile>
          </>
        )}
      </Toggle>
    );
  }
}

export default Starred;
