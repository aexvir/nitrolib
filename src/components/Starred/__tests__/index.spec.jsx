// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Starred from "..";

describe("#Starred", () => {
  test("render open", () => {
    const wrapper = shallow(
      <Starred positionMenuDesktop={0} onSelectStarred={jest.fn()} positionMenuTablet={0} />,
    );

    expect(wrapper.prop("children")({ open: true, onToggle: jest.fn() })).toMatchSnapshot();
  });

  test("render closed", () => {
    const wrapper = shallow(
      <Starred positionMenuDesktop={0} onSelectStarred={jest.fn()} positionMenuTablet={0} />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
