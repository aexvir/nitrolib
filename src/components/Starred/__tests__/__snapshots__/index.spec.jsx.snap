// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Starred render closed 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`#Starred render open 1`] = `
<Toggle>
  [Function]
</Toggle>
`;
